### PR TITLE
[SPARK-15767][ML][SparkR] Decision Tree wrapper in SparkR

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
support decision tree in R

## How was this patch tested?
added tests

### CODE REVIEW

**Comment by zhengruifeng** [comment]
Jenkins, please retest this please

**Comment by zhengruifeng** [comment]
@felixcheung I send this PR following your implementation of `RandomForest`. Could you help reviewing this?

File: R/pkg/R/mllib_tree.R [hunk 1]
```diff
@@ -45,6 +45,20 @@ setClass("RandomForestRegressionModel", representation(jobj = "jobj"))
 #' @note RandomForestClassificationModel since 2.1.0
 setClass("RandomForestClassificationModel", representation(jobj = "jobj"))
 
+#' S4 class that represents a DecisionTreeRegressionModel
+#'
+#' @param jobj a Java object reference to the backing Scala DecisionTreeRegressionModel
+#' @export
+#' @note DecisionTreeRegressionModel since 2.2.0
```
Comment (felixcheung) [inline R/pkg/R/mllib_tree.R]: 2.3.0 please

File: R/pkg/R/mllib_utils.R [hunk 2]
```diff
@@ -38,6 +38,7 @@
 #' @seealso \link{spark.lda}, \link{spark.logit},
 #' @seealso \link{spark.mlp}, \link{spark.naiveBayes},
 #' @seealso \link{spark.randomForest}, \link{spark.survreg}, \link{spark.svmLinear},
+#' @seealso \link{spark.decisionTree},
```
Comment (felixcheung) [inline R/pkg/R/mllib_utils.R]: please keep this sorted

File: R/pkg/R/mllib_utils.R [hunk 3]
```diff
@@ -52,7 +53,8 @@ NULL
 #' @seealso \link{spark.gbt}, \link{spark.glm}, \link{glm}, \link{spark.isoreg},
 #' @seealso \link{spark.kmeans},
 #' @seealso \link{spark.logit}, \link{spark.mlp}, \link{spark.naiveBayes},
-#' @seealso \link{spark.randomForest}, \link{spark.survreg}, \link{spark.svmLinear}
+#' @seealso \link{spark.randomForest}, \link{spark.survreg}, \link{spark.svmLinear},
+#' @seealso \link{spark.decisionTree}
```
Comment (felixcheung) [inline R/pkg/R/mllib_utils.R]: keep this sorted

File: R/pkg/inst/tests/testthat/test_mllib_tree.R [hunk 4]
```diff
@@ -209,4 +209,91 @@ test_that("spark.randomForest", {
   expect_equal(summary(model)$numFeatures, 4)
 })
 
+
```
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_mllib_tree.R]: nit: remove double empty line

File: R/pkg/R/mllib_tree.R (hunk 1 above)
Comment (felixcheung) [inline R/pkg/R/mllib_tree.R]: and all other instances...

**Comment by zhengruifeng** [comment]
@felixcheung Updated. Thanks for your reviewing!

File: R/pkg/R/mllib_tree.R [hunk 5]
```diff
@@ -499,3 +543,199 @@ setMethod("write.ml", signature(object = "RandomForestClassificationModel", path
...
+#'                 "entropy" and "gini", default is "gini".
+#' @param seed integer seed for random number generation.
+#' @param minInstancesPerNode Minimum number of instances each child must have after split.
+#' @param minInfoGain Minimum information gain for a split to be considered at a tree node.
+#' @param checkpointInterval Param for set checkpoint interval (>= 1) or disable checkpoint (-1).
+#' @param maxMemoryInMB Maximum memory in MB allocated to histogram aggregation.
+#' @param cacheNodeIds If FALSE, the algorithm will pass trees to executors to match instances with
+#'                     nodes. If TRUE, the algorithm will cache node IDs for each instance. Caching
+#'                     can speed up training of deeper trees. Users can set how often should the
+#'                     cache be checkpointed or disable it by setting checkpointInterval.
```
Comment (bdwyer2) [inline R/pkg/R/mllib_tree.R]: This is kind of confusing
>Users can set how often should the cache be checkpointed

Comment (felixcheung) [inline R/pkg/R/mllib_tree.R]: wording can be improved a bit I guess but this matches the Scaladoc...

File: R/pkg/R/mllib_tree.R [hunk 6]
```diff
@@ -499,3 +543,199 @@ setMethod("write.ml", signature(object = "RandomForestClassificationModel", path
...
+#' t <- as.data.frame(Titanic)
+#' df <- createDataFrame(t)
+#' model <- spark.decisionTree(df, Survived ~ Freq + Age, "classification")
+#' }
+#' @note spark.decisionTree since 2.3.0
+setMethod("spark.decisionTree", signature(data = "SparkDataFrame", formula = "formula"),
+          function(data, formula, type = c("regression", "classification"),
+                   maxDepth = 5, maxBins = 32, impurity = NULL, seed = NULL,
+                   minInstancesPerNode = 1, minInfoGain = 0.0, checkpointInterval = 10,
+                   maxMemoryInMB = 256, cacheNodeIds = FALSE) {
```
Comment (felixcheung) [inline R/pkg/R/mllib_tree.R]: consider adding `thresholds` parameter - possibly as a follow up PR.

**Review by felixcheung — APPROVED** [review]
LGTM.

**Comment by felixcheung** [comment]
any more comment?

**Comment by felixcheung** [comment]
merged to master. thanks!

**Pull request closed without merging**